### PR TITLE
Auto recovery

### LINKED ISSUE #104: Recovery from crashes
With #78 now done we have a pretty comprehensive way to doo full integrity checks on the entire database and potentially "recovery" from corruption.

Just some further thoughts on this as I was poking back around at the original papaer the [Riak's docs](https://docs.riak.com/riak/kv/2.1.4/setup/planning/backend/bitcask/index.html#bitcask-s-strengths)

Specifically:

> * **Fast, bounded crash recovery**
>
> Crash recovery is easy and fast with Bitcask because Bitcask files are append only and write once. The only items that may be lost are partially written records at the tail of the last file that was opened for writes. Recovery operations need to review only the last record or two written and verify CRC data to ensure that the data is consistent.

This leads me to believe we _should_ be able to automatically reco ver from crashes by inspecting the last N items in the "active datafile" for correctness/integrity? We could automatically recover from such failures and put the "bad data" aside for the operator to inspect. Thoughts?

### LINKED ISSUE #143: Corrupt index
For whatever reason, when a program crash occurs, the `index` file becomes corrupt and any attempt to reconnect to the bitcask datastore hangs until the index is removed, at which point, obvious problems ensue.

Is there a way to cleanly rebuild the index on startup?

### PR DESCRIPTION
**Pull request by jsign**
This PR implements a new feature: Auto Recovery.
For more context, see #104.

In short, if the database is opened with `WithAutoRecovery(true)` (`false` by default), opening a database check if the _latest_ datafile is corrupted. If it isn't the case, it's a noop. If it is, it recovers as much of the file as it can, and deletes the index which will be autogenerated again.

I made explicit in the docs that this new option must be run considering that the user has done proper backups of the data, if interested. I avoided always doing an automatic backup since that might create further assumptions if that's wanted. Of course, this can be reconsidered if we realize it may be safe otherwise.

The code needs some extra refactor to avoid some code duplication on the `recover` command. But that that is left as a further minor ToDo to let this long-wanted feature land. It may be nice to create an issue and tag it as `good-first-issue` or similar.

I included a test that covers all this is working correctly.

Fixes #104 
Fixes #143

### CODE REVIEW

**Comment by jsign** [comment]
All tests are green unless 1.13 on windows (which is OK on 1.12).
I don't have available a windows machine to check what's happening, but from the failure, I'd guess there's some problem with removing the stalled-index file in the recovery process.

**Comment by prologic** [comment]
Same here :)

File: internal/data/recover.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	"github.com/prologic/bitcask/internal"
+	"github.com/prologic/bitcask/internal/config"
+	"github.com/prologic/bitcask/internal/data/codec"
+)
+
+// CheckAndRecover checks and recovers the last datafile.
+// If the datafile isn't corrupted, this is a noop. If it is,
+// the longest non-corrupted prefix will be kept and the rest
+// will be *deleted*. Also, the index file is also *delete* which
```
Comment (prologic) [inline internal/data/recover.go:17]:
```suggestion
// will be *deleted*. Also, the index file is also *deleted* which
```

File: internal/data/recover.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		return nil
+	}
+	f := dfs[len(dfs)-1]
+	recovered, err := recoverDatafile(f, cfg)
+	if err != nil {
+		return fmt.Errorf("recovering data file")
+	}
+	if recovered {
+		if err := os.Remove(filepath.Join(path, "index")); err != nil {
+			return fmt.Errorf("deleting the index on recovery: %s", err)
```
Comment (prologic) [inline internal/data/recover.go:34]:
```suggestion
			return fmt.Errorf("error deleting the index on recovery: %s", err)
```

File: internal/utils.go [hunk 3]
```diff
@@ -33,7 +33,8 @@ func DirSize(path string) (int64, error) {
 
 // GetDatafiles returns a list of all data files stored in the database path
 // given by `path`. All datafiles are identified by the the glob `*.data` and
-// the basename is represented by an monotomic increasing integer.
+// the basename is represented by an monotonic increasing integer.
```
Comment (prologic) [inline internal/utils.go:36]:
```suggestion
// the basename is represented by a monotonic increasing integer.
```

**Comment by prologic** [comment]
Hmm your test is fialing on Windows?

**Comment by jsign** [comment]
> Hmm your test is fialing on Windows?

Yep, mentioned that on a previous comment.

**Comment by prologic** [comment]
> Yep, mentioned that on a previous comment.

Are you going to fix that on this PR, skip the tests for Windows or something else? :)

**Comment by jsign** [comment]
> > Yep, mentioned that on a previous comment.
> 
> Are you going to fix that on this PR, skip the tests for Windows or something else? :)

Voted for skipping only in Windows since I don't have a Windows PC to see what's really the problem. I'll create a `help-wanted` issue if someone with a Windows PC wants to take a look.

**Pull request merged**